Windows: Acquire bootstrap resources

johnwparent opened the pull request:
Closes https://github.com/spack/spack/pull/39729

File and GPG cannot be built on Windows, nor can Spack cross compile them, but these utilities are needed to run a number of Spack utilities. This PR:

- Adds a mechanism to download pre-built binaries by name (vs. the typical route of by-spec in `bootstrap/`)
- Updates low-level gpg/file-utility access to automatically retrieve these resources when needed
- Because these bootstrapping actions are triggered by accessing the utilities (vs. the normal procedure of bootstrap installing gpg up front), this new bootstrapping mechanism is split into a separate module (`binary_resource`) to avoid cyclic import
- It wasn't strictly necessary to this PR, but those functions in `llnl.util.filesystem` that need `spack.util` were also split off into their own module

johnwparent committed:
Windows: Bootstrap Resources

Windows systems are frequently bereft of utilities
imperative to Spack's operations
These are typically either available on systems by
default on other platforms, or are installable by
Spack directly.
On Windows these utilities must be cross compiled
from unix. To provide these for Spack we introduce
bootstrap resources, precompiled by a GHA
workflow, and acquired by Spack to resolve
bootstrap requirements.

Currently supports:

* gpg
* file

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -53,8 +57,12 @@
 import spack.util.url
 import spack.version
 
+import spack.util.web as sw
```
Typically this is imported `as web_util`

johnwparent replied:
Sounds good!

scheibelp commented on share/spack/resource/github-actions-v0.1/layout.yaml:
```diff
@@ -0,0 +1,7 @@
+resource:
```
Is this schema defined by github somewhere?

Either way, I'm not clear why these details aren't merged into `resource.yaml`.

If it's not defined elsewhere, then this structure should be codified in a schema.

johnwparent replied:
This is actually more or less a very simplified version of the scheme PyPa uses (I pulled the resource CD concept from a GH Pages hosted PyPa alternative I have setup).

> I'm not clear why these details aren't merged into resource.yaml

I was following the pattern for other, similar configs in Spack (bootstrapping cache mostly), which follow this approach. I can just fold it into `resource.yaml`. 

> then this structure should be codified in a schema

On it! (Following advice from comment above I'll add it to the `resource` schema

scheibelp commented on etc/spack/defaults/windows/resource.yaml:
```diff
@@ -0,0 +1,8 @@
+bootstrap-resource:
+  # If set to false Spack will not fetch bootstrap resources but instead
+  # proceed without fetching said resources
+  enable: true
```
Why allow this to be disabled?

johnwparent replied:
I thought some users might take objection with the inability to prevent the download of semi-arbitrary binaries, particularly if they want to vendor those binaries themselves and it could cause issue for non-networked machines.
That was 100% of my reasoning, if you don't think it's a relevant concern, I'll pull it out no problem.

scheibelp replied:
Interesting point, I like the concept about being explicit, although I wonder if we ought to collect these sorts of concerns into a higher-level dont-download-things variable (e.g. which we can use to disable a centralized download function that we use) - vs. distributing these sorts of concerns across multiple configs (if someone has the concern you express, they'd probably be frustrated to disable it here and then get a download somewhere else.

That shouldn't hold this up, but I'd like to discuss how to avoid a dispersion of these sorts of concerns.

> for non-networked machines.

could we have a prior check: look for `file` and only download it if it isn't present?

johnwparent replied:
> if we ought to collect these sorts of concerns into a higher-level dont-download-things variable 

I'm in support of any degree of centralization we can pull from the configs. I agree that it might be out of scope for this MR though.

> could we have a prior check: look for `file` and only download it if it isn't present?

Great idea, will implement.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -493,6 +513,49 @@ def _add_externals_if_missing() -> None:
+    for resource in resources:
+        if resource.acquire_resource():
+            insert_resource_into_environment(name)
+            return
```
It seems like there's a disconnect: you call

```
    ensure_resource("file")
    ensure_resource("gpg")
```

but in this context, the resource is the `layout.yaml` that stores multiple sources, which all appear to be unpacked here.

This is partially exacerbated by reuse of the term `resource`: i.e. there's `BootstrapResource`, the existing `Resource` object in Spack, and calling the the endpoints `resources` in this function (can you create more-distinct names?).

johnwparent replied:
You're right there is. This method needs a refactor, the interface of calling `ensure_resource` is what I had intended, but the resource gathering doesn't consider that. 
I'll refactor (and rename)

scheibelp reviewed (changes requested):
I have some questions and a few higher-level concerns (although the overall strategy looks good).

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -569,12 +650,29 @@ def ensure_winsdk_external_or_raise() -> None:
-        ensure_gpg_in_path_or_raise()
+    if IS_WINDOWS:
+        ensure_win_resources()
+        ensure_file_in_path_or_raise()
```
Is this redundant with `ensure_resource("file")`?

johnwparent replied:
Yes, as is the GPG check below. Removed the file check entirely and then made the gpg check unix only.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -468,6 +491,64 @@ def _add_externals_if_missing() -> None:
+
+
+def ensure_resource(name):
+    """Acquires resource from configured sources"""
```
`insert_resource_into_environment` is Windows-specific, so the name or docstring of this function should mention that it is Windows-specific.

johnwparent replied:
Makes sense, pushed.

scheibelp replied:
(minor) I was assuming a change addressing this would change the name or docstring of this function - if you handled this another way, could you point me to where?

johnwparent replied:
~Oh I think I just missed this comment, sorry!~
Actually I think I renamed another method to `win_insert_resource_into_environment` (the method that actually does the inserting into the environment) and I already have a method called `win_ensure_resources` that calls this.
Renamed this to `win_ensure_or_acquire_resource`

scheibelp commented on etc/spack/defaults/windows/bootstrap-resource.yaml:
```diff
@@ -0,0 +1,16 @@
+      providers:
+        - name: Github Actions
+          endpoint: https://spack.github.io/windows-bootstrap-resources/resources/gpg4win/gpg4win.tar.gz
+          sha256: notasha
```
This is a strange placeholder: should we in fact be including the checksums here and verifying them?

johnwparent replied:
Yes, I'm just in the process of updating the workflow that produces them so I don't have actual sha's right now. I thought this would be obvious enough to make sure we won't forget before a merge.

johnwparent replied:
I have a sha for gpg, just not file, ~not sure what changed there~ I updated the mingw image and now I'm suddenly missing a dep. Should be resolved before we're ready to hit merge here.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -468,6 +491,64 @@ def _add_externals_if_missing() -> None:
+        )
+    resources = spack.config.get("bootstrap-resource:resources")
+    providers = resources.get(name)
+    for provider in providers:
```
I'm not clear how this works: the file layout is

```
  resources:
    gpg:
      providers:
        - name: Github Actions
          endpoint: https://spack.github.io/windows-bootstrap-resources/resources/gpg4win/gpg4win.tar.gz
          sha256: notasha
```

so if you `.get(...:resources)` and index by `name`, you should end up with a dict like `{providers: []}`

johnwparent replied:
You are correct, made a mistake copying it over from my test script.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -69,6 +74,24 @@
+def _bootstrap_resources_enabled() -> bool:
+    """Returns whether fetching binary resoures for bootstrapping
+    is enabled"""
+    return spack.config.get("bootstrap-resource:enable")
```
Now that we check ahead of time and bail if the resource is available in the PATH, I think this could either be:

* removed
* Or defer to `bootstrap:enable`

Also, I'm not clear if there is a benefit to this one-liner function (the `config.get` path is clear enough to appear by itself).

johnwparent replied:
The resource fetching and the bootstrapping process as a whole introduce enough disparity w.r.t. expectations of the current bootstrapping process that I feel like we still need an explicit setting. Particularly because there is no way to complete the resource fetching without a network connection, and we're never building from source locally, both of which are not true for bootstrapping itself (I assume bootstrapping can pull from local source/binary caches the same way standard installs can).

We are checking for the existence of the binaries we need now, but if we don't find them, I don't think that's enough reason to not provide the user the ability to prefer to fail at that point rather than download binaries.

Maybe I'm paranoid about a use case that doesn't exist, but I'd rather have an explicit "failed to bootstrap Spack because X was not found on system and resource fetching has been disabled" failure rather than a download failure (particularly in systems with restrictive network access).

One liner comment - agreed, will put it back in the method.

scheibelp replied:
> I'd rather have an explicit "failed to bootstrap Spack because X was not found on system and resource fetching has been disabled" failure rather than a download failure (particularly in systems with restrictive network access).

I agree - it looks like you achieved this in `ensure_resource`

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -69,6 +74,24 @@
 
+def _windows_resource_root() -> pathlib.Path:
+    """Returns the root of the Windows resources required for bootstrapping"""
+    return pathlib.Path(root_path()) / "windows-resources"
```
IMO this would be clearer to organize alongside `store_path` etc. in `bootstrap/config`, as a public function

johnwparent replied:
Pushed

scheibelp commented on lib/spack/spack/schema/bootstrap_resource.py:
```diff
@@ -0,0 +1,53 @@
+        "additionalProperties": False,
+        "properties": {"enable": {"type": "boolean"}, "resources": resource_entry},
+    }
+}
```
(minor) IMO this full decomposition makes this a bit harder to read (vs. if you were to "collapse" most of it together): that tends to be useful for splitting off larger sections (or e.g. `resource_provider`), but as far as a dict/dict/list/... it is easier to see that structure when combined.

johnwparent replied:
I felt the exact opposite while writing the schema :rofl: lol.
I'll consolidate it as requested.

scheibelp commented on etc/spack/defaults/windows/bootstrap-resource.yaml:
```diff
@@ -0,0 +1,16 @@
+  resources:
+    gpg:
+      providers:
+        - name: Github Actions
```
(question) was Github Actions generating these? What action triggers that? Either way, I think this could be clearer, since it would be the github actions of another repo (https://github.com/spack/windows-bootstrap-resources) that generates these, so it may be more informative to mention that this comes from a Github-hosted repo.

johnwparent replied:
A github actions workflow is yes, on push to develop in that repo you linked, the binaries are regenerated by a GHA job and the old binaries are replaced on the pages instance serving them.


The workflow in that repo is:
- Make changes to a dockerfile for mingw-w64, gpg, or file.
- Commit changes to branch, push branch to resource repo, make PR of that branch.
- GHA will build all the dockerfiles and verify they succeed and produce binaries
- Merge to develop
- Develop builds latest dockerfiles, extracts binaries, then checks out the pages branch, deletes the current binaries, checks the new binaries in, and then updates the html in the pages branch to serve the new binaries.

johnwparent replied:
Renamed to "Spack Bootstrap Resource Repo" and since the url beneath includes the repo I think that covers our bases.

My original intent was to keep things in line with the bootstrap mirrors named things like `Github-Actions-v1`

scheibelp commented on etc/spack/defaults/windows/bootstrap-resource.yaml:
```diff
@@ -0,0 +1,16 @@
+bootstrap-resource:
```
As I compare this and `bootstrap.yaml`, I think it would actually be worthwhile to merge this into the `bootstrap` schema as a subsection - does that seem agreeable?

johnwparent replied:
If that won't make it too busy I don't see why not.

scheibelp reviewed (changes requested):
Largest request is to merge `bootstrap-resources.yaml` into `bootstrap.yaml`. Let me know if that or any of the other requests seem off.

johnwparent committed:
Update now that binaries are up

johnwparent committed:
add bootstrap location to path for exe search

johnwparent committed:
Ensure acquired bootstrap resources can actually be used

alalazo commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -251,6 +251,11 @@ def path_contains_subdirectory(path, root):
 @memoized
 def file_command(*args):
     """Creates entry point to `file` system command with provided arguments"""
+    if sys.platform == "win32":
+        import spack.bootstrap
+
+        with spack.bootstrap.ensure_bootstrap_configuration():
+            spack.bootstrap.ensure_resource("file")
```
Please, let's not add more `llnl->spack->llnl` circular deps

johnwparent replied:
Moved everything in `llnl.util.filesystem` that requires an import from core spack to a core spack module.

alalazo reviewed (changes requested):
Can we have a high level description of the PR (what is it about, how it performs operations at a high level etc.)?

scheibelp commented on lib/spack/spack/schema/bootstrap.py:
```diff
@@ -5,6 +5,38 @@
+    },
+}
+
+resource_entry = {
```
(minor) this name implies to me it is for describing a single resource, but in fact it covers all of them: I'd recommend renaming to `resources_section` or something like that.

johnwparent replied:
Good point, thanks!

scheibelp commented on lib/spack/spack/config.py:
```diff
@@ -1553,6 +1553,8 @@ def ensure_latest_format_fn(section: str) -> Callable[[YamlConfigDict], bool]:
     # The line below is based on the fact that every module we need
     # is already imported at the top level
+    # allow for hyphenated section naming
+    section = section.replace("-", "_")
```
I think this is no longer required (on one hand could be useful for future config sections, but one thing I avoided getting into based on other changes was having to look into whether this had any unexpected ramifications, so IMO it would be good to remove here).

johnwparent replied:
I had the same thought it could be useful for future cases with hyphenated names but that's a fair point, will remove!

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -569,11 +640,20 @@ def ensure_winsdk_external_or_raise() -> None:
+    if IS_WINDOWS:
+        # Windows also requires gpg (as in the check below)
+        # but that is resolved by this call
+        ensure_win_resources()
```
The way you have factored this such that accessing the resource directly always bootstraps it, is this part still necessary? I think this could be removed (and I think that will still be true after addressing https://github.com/spack/spack/pull/41810#discussion_r1629121088, but you might want to address that comment first in case I'm wrong).

johnwparent replied:
I was hoping to also fetch these resources with the `spack bootstrap now` command which invokes this `ensure_core_dependencies` method.
So necessary? No. That said I think we should leave it. We won't ever be "rebootstrapping" any of these so I don't think it hurts to have here?

scheibelp replied:
> I don't think it hurts to have here?

I don't think it will cause problems directly, but my concern is more that if we set up multiple pathways for something to be addressed, it will not be clear which one is the required one. Someone could come along later, notice both of these, then remove the other. In general, if there is a 100% solution and a less-than-100% pathway that also exists, I prefer to remove the latter.

johnwparent replied:
To be clear, I am not opposed to removing this, but I would like to understand why this is a different case from something like GPG  or Clingo, where we do have multiple paths to bootstrap them so I can make the right call in the future. For those two, we have the `bootstrap now` approach, which invokes the method currently in question, and then the "normal spack flow" approach, where a spack command encounters the need for clingo/gpg and Spack calls into the bootstrap at that point. They end up calling the same methods under the hood, but so does this.

scheibelp replied:
I didn't see there were multiple pathways for Clingo. IMO Clingo bootstrapping is distinct in that it could take a significant amount of time: one of the automated fallback bootstrapping mechanisms is to install from source; this is true to a lesser extent for GPG on Unix as well. On Windows, these bootstrapping operations are always fast (because downloading the binaries is the only option).

For that reason, it's ideal to bootstrap Clingo/GPG up front (in violation of the general principle of only having one pathway for things).

scheibelp reviewed (changes requested):
This looks good at a high level, the only requests I have left are all minor.

johnwparent committed:
WIP: test new images in prod because staging won't work

win+9

scheibelp commented on lib/spack/spack/config.py:
```diff
@@ -1553,6 +1553,7 @@ def ensure_latest_format_fn(section: str) -> Callable[[YamlConfigDict], bool]:
     """
     # The line below is based on the fact that every module we need
     # is already imported at the top level
+    # allow for hyphenated section naming
```
vestigial comment

scheibelp commented on lib/spack/spack/util/system.py:
```diff
@@ -0,0 +1,93 @@
+
+"""
+Utility methods for interfacing with various system utilities
+"""
```
IMO `system` is too generic a name. We call our Spack augmented YAML handler `spack_yaml`, and I think that would be appropriate to emulate here, like `spack _filesystem`.

scheibelp commented on lib/spack/spack/util/system.py:
```diff
@@ -0,0 +1,93 @@
+def file_command(*args):
+    """Creates entry point to `file` system command with provided arguments"""
+    if sys.platform == "win32":
+        import spack.bootstrap
```
You can promote this to a module-level import now (or at least, I think that was the hope for pulling these functions into another module).

johnwparent replied:
I tried that, it still causes circular imports due to the Environment module.

scheibelp replied:
(just adding circular import error, from perspective of `system.py`)

```
  (system.py)
    import spack.bootstrap
  File "C:\spack_win\spack\lib\spack\spack\bootstrap\__init__.py", line 8, in <module>
    from .core import (
  File "C:\spack_win\spack\lib\spack\spack\bootstrap\core.py", line 51, in <module>
    import spack.user_environment
  File "C:\spack_win\spack\lib\spack\spack\user_environment.py", line 9, in <module>
    import spack.build_environment
  File "C:\spack_win\spack\lib\spack\spack\build_environment.py", line 56, in <module>
    import spack.build_systems.cmake
  File "C:\spack_win\spack\lib\spack\spack\build_systems\cmake.py", line 19, in <module>
    import spack.package_base
  File "C:\spack_win\spack\lib\spack\spack\package_base.py", line 57, in <module>
    from spack.filesystem_view import YamlFilesystemView
  File "C:\spack_win\spack\lib\spack\spack\filesystem_view.py", line 38, in <module>
    import spack.relocate
  File "C:\spack_win\spack\lib\spack\spack\relocate.py", line 28, in <module>
    import spack.util.system as ssys
```

scheibelp commented on lib/spack/spack/util/gpg.py:
```diff
@@ -62,7 +63,10 @@ def init(gnupghome=None, force=False):
+        if sys.platform == "win32":
+            spack.bootstrap.win_ensure_or_acquire_resource("gpg")
+        else:
+            spack.bootstrap.ensure_gpg_in_path_or_raise()
```
(minor) I'm thinking it might be nice to move this `if` inside `spack.bootstrap` and have a higher-level `give-me-gpg` function (apologies if this has been this way for a while and I'm just now asking about it - although I think this became clearer to me as you reorganized/improved things).

johnwparent replied:
Makes a lot of sense, will address.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -468,6 +479,66 @@ def _add_externals_if_missing() -> None:
+        resource = spack.resource.Resource(name, fetcher, destination=stage.path, placement=None)
+        self.stage = spack.stage.ResourceStage(
+            fetcher, stage, resource, path=str(windows_resource_root()), keep=True
+        )
```
IMO it would be more idiomatic for this to:

* Create a stage without specifying a path (thereby downloading into the stage's tmp dir)
* From in the context manager, copy into `windows_resource_root`
* Allow `keep=False`, so the stage is cleaned up

vs. treating the `windows_resource_root` as the Stage root. Does that seem agreeable?

johnwparent replied:
Yeah that makes a lot of sense, will refactor, thanks, good suggestion, this did feel awkward.

johnwparent committed:
wip

johnwparent committed:
rename spack system to spack filesystem

johnwparent committed:
rework stage component of bootstrap resource

johnwparent committed:
style

johnwparent committed:
Refactor resource yaml scheme

johnwparent committed:
Refactor binary resource fetching into operation independent of bootstrap

scheibelp commented on lib/spack/spack/schema/resource.py:
```diff
@@ -0,0 +1,50 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
update to 2024

Can we call this something else to distinguish it from the `resource` directive?

`binary_resource`/`bootstrap_resource`. I can't think of a name that isn't two part. FWIW I don't think there's a need to convert from `-` to `_` (e.g. we can force users to `spack config blame bootstrap_resource` vs. `... bootstrap-resource`).

scheibelp commented on lib/spack/spack/util/binary_resource.py:
```diff
@@ -0,0 +1,88 @@
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Interface for acquiring Spack binary resources required for bootstrapping"""
```
Suggested reword:

> Interface for bootstrapping needed resources for running of Spack; unlike bootstrap.py (the typical pathway) this does not refer to utilities with Specs, but rather by an ID that refers to the needed binary. This is currently only needed on Windows.

scheibelp commented on lib/spack/spack/util/spack_filesystem.py:
```diff
@@ -0,0 +1,92 @@
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""
+Utility methods for interfacing with various system utilities
```
Suggested reword:

> Utilities for interacting with files, like those in llnl.util.filesystem, but which require logic from spack.util

scheibelp reviewed (changes requested):
A couple doc requests, and a request to rename the new schema.

johnwparent committed:
Review + completion

johnwparent commented:
@scheibelp this is ready for another round of review

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -569,6 +573,11 @@ def ensure_winsdk_external_or_raise() -> None:
 
+def ensure_win_resources() -> None:
+    spack.util.binary_resource.win_ensure_or_acquire_resource("file")
+    spack.util.binary_resource.win_ensure_or_acquire_resource("gpg")
```
Is this used at the moment? Or were you planning on keeping it regardless?

(side note: I'm surprised all the imports added here are still needed: I would have assumed the style checker would complain now that most of them are no longer referenced in this diff)

johnwparent replied:
It is unused. That said I was planning on keeping it just so we have some reflection of this "bootstrapping" procedure in the bootstrapping module, mostly so future devs looking for this behavior in the bootstrapping module (which, since we named the config `bootstrap_resources`, seems likely) find this. 

I'm not overly tied to having it there, if you want we can remove.

johnwparent replied:
> (side note: I'm surprised all the imports added here are still needed: I would have assumed the style checker would complain now that most of them are no longer referenced in this diff)

I thought it would too. Weirdly, despite not considering them unused, none of the style checks griped when I took those imports out, so I'll just do that.

scheibelp replied:
> I was planning on keeping it just so we have some reflection of this "bootstrapping" procedure in the bootstrapping module

That works with me!

scheibelp reviewed (commented):
I had a question about some code that looks vestigial. If you intended to keep it, that should be fine, but it looks like all changes could be removed from `bootstrap/core`. Otherwise this looks good!

johnwparent committed:
rm dangling imports

scheibelp commented on lib/spack/spack/util/binary_resource.py:
```diff
@@ -0,0 +1,94 @@
+        tty.debug(f"Resource {name} already available on system path at: {cmd.path}")
+        win_insert_resource_into_environment(name)
+        return
+    if not cmd and not spack.config.get("resource:enable"):
```
`resource` was renamed to `bootstrap_resource`.

scheibelp reviewed (changes requested):
Apologies: I just noticed that with recent changes, there seems to be a reference error. With that in mind, it would be useful to have a unit test for `win_ensure_or_acquire_resource` (perhaps one that mocks `BinaryResource.acquire_resource`)

johnwparent committed:
Repair invalid reference

johnwparent committed:
Code structure improvements

johnwparent committed:
Add test

johnwparent committed:
style

johnwparent committed:
Env sandboxing

johnwparent committed:
Invoke file directly

johnwparent committed:
Fix tests

johnwparent committed:
Style

johnwparent commented:
@alalazo see updated PR description

scheibelp commented on lib/spack/spack/util/binary_resource.py:
```diff
@@ -0,0 +1,100 @@
+        resource = spack.resource.Resource(
+            name, fetcher, destination=self.resource_subdir, placement=None
+        )
+        self.stage = spack.stage.ResourceStage(fetcher, stage, resource, keep=False)
```
(question) I think we spoke about using `Stage` directly vs. constructing a `ResourceStage`, did that not work out?

scheibelp commented on lib/spack/spack/test/util/binary_resource.py:
```diff
@@ -0,0 +1,90 @@
+
+@pytest.fixture
+def on_system(working_env):
+    """Fixture explicityly adding binary resource to PATH"""
```
(typo) explicityly

scheibelp commented on lib/spack/spack/test/util/binary_resource.py:
```diff
@@ -0,0 +1,90 @@
+
+@pytest.fixture
+def tar_dir():
+    return os.path.dirname(spack.util.executable.which("tar").path)
```
I think it would be more-straightforward to name this fixture `add_tar_to_path` (IMO it could also handle modifying/appending to `PATH` - it can take `no_system` as a param to enforce fixture ordering): it was initially confusing what this was for until I read through how it was used.

scheibelp commented on lib/spack/spack/test/util/binary_resource.py:
```diff
@@ -0,0 +1,90 @@
+def test_ensure_or_acquire_resource_on_system(
+    mock_binary_resource_root, mutable_config, on_system
+):
+    # if resource not on system, ensure the resource acquisition fails with an exception
```
If I understand, you meant "I've set this up so that if `file` is not on the system, that this will fail". FWIW I think the next test covers this sufficiently, and it does not depend on the availability of `file`. If you want to cover this, I'd recommend updating the next test to add a fake `file` binary and then re-execute after the first (expected) failure.

johnwparent replied:
My intent here was to explicitly test the short circuiting logic I have that avoids fetching any resources when file _is_ present on the system, where the failure here would indicate an issue with that logic. 
I'll consolidate as suggested

scheibelp commented on lib/spack/spack/test/util/binary_resource.py:
```diff
@@ -0,0 +1,90 @@
+
+
+@pytest.mark.skipif(not spack.util.executable.which("tar"), reason="tar required; not available")
+def test_ensure_or_acquire_acquire_resource(mock_binary_resource_root, config, tar_dir, no_system):
```
I think this should mention that the test config stores a file URL to a test resource.

scheibelp commented on lib/spack/spack/test/util/binary_resource.py:
```diff
@@ -0,0 +1,90 @@
+    env.apply_modifications()
+
+
+def test_ensure_or_acquire_no_acquire(mock_binary_resource_root, config, no_system, monkeypatch):
```
If I understand, this and the next test mock acquisition at two different levels:

* In this case, mocking `acquire_resource` (such that issues with fetching etc. would not cause this test to fail
* In the next case, allowing `acquire_resource` to run in full

This type of coverage is rare for Spack tests. Typically I'd recommend only one (in this case, the latter); if you keep both around, I think a different name would be clearer, because this *is* calling `acquire`, and in fact we are making sure this pathway calls it. I might call it `test_try_acquire` i.e. "make sure we at least attempt to acquire a resource when we find it is not available".

johnwparent replied:
Will just remove this test.

scheibelp reviewed (changes requested):
I have a few requests about test organization, and a question about the use of `ResourceStage` in `BinaryResource`.

johnwparent committed:
review

johnwparent committed:
style

johnwparent committed:
Style

haampie commented:
Why don't you implement it as a package: `spack install file-windows-bootstrap` or so.

alalazo reviewed (commented):
I basically have the same question as @haampie: can't we have a specific package to install binaries without additional boilerplate code + additional YAML files?

Also, side question: what is currently preventing us from building those packages from sources?

scheibelp commented:
> can't we have a specific package to install binaries without additional boilerplate code + additional YAML files?

That would work, but I'm not sure if I understand what you mean by "boilerplate": I would interpret it to mean "code that wouldn't need to exist if we followed the other approach"; note that if we followed this other approach, either:

1. new code would need to be introduced to address the 2-layer bootstrapping described in https://github.com/spack/spack/pull/41810#issue-2052823805
2. or... this would want to execute bootstrapping upfront as before, and be more careful to do it in more places (to avoid ever calling the functions that need `file`/`gpg` without having bootstrapped those tools)

Overall I feel like 

* The approach in this PR has an advantage over [2]
* I'm not clear how to compare this written code with unwritten code that supports [1]: overall there's ~150 lines of code (`bootstrap_resource.py` schema, `binary_resource.py`) essential to the current approach; IMO that's worth it for something more direct, by that I mean:
  * Addressing this with packages creates some awkwardness: these packages don't have real versions, and are irrelevant outside of usage within core spack (presumably we'd want a separate package repo for them)
  * New resources would be faster to add vs. with a package-based approach (in that sense the PR in it's current form avoids some boilerplate code that comes w/ using packages.

So I think this is good as is. If you disagree, please add a request-for-changes (@alalazo I realize you have a request already, but the requests from that review have been addressed).

alalazo reviewed (changes requested):
> new code would need to be introduced to address the 2-layer bootstrapping described in https://github.com/spack/spack/pull/41810#issue-2052823805

Sorry, I don't understand what you mean by 2-layer bootstrapping. 

> Addressing this with packages creates some awkwardness: these packages don't have real versions, and are irrelevant outside of usage within core spack (presumably we'd want a separate package repo for them)

This is how we do bootstrapping since we started bootstrapping clingo. We have a package `clingo-bootstrap` which is just `clingo` with some options/tweaks that are relevant to Spack only. We also have a `@spack` version, since so far we are still using a pinned commit. I'm not saying this is not awkward in some sense, I'm just saying that this PR introduces yet another way of doing things + new awkwardness, which should be avoided if possible. 

Also, doesn't Windows bootstrap clingo from sources, with the usual mechanism? If that's the case Windows would need 2 different mechanisms and code paths for two different dependencies of Spack. This would be inconsistent.

> Also, side question: what is currently preventing us from building those packages from sources?

I don't see a clear answer to this anywhere. I don't understand what "Can't be built on Windows" means, if we install binaries for the thing that cannot be built.

johnwparent commented:
@alalazo 

> What does it mean "cannot be built on windows", how did obtain binaries then?

> what is currently preventing us from building those packages from sources?

Neither File, nor GnuPG have build systems that are compatible with being built natively on Windows. They are autotools/make only projects. Normally, like I did for Lua, I would at least attempt to create a CMake system since both projects don't change often, however neither piece of software is capable of compiling natively on Windows due to their reliance on posix libraries/headers. So we have to cross compile. Spack can't do that. We could add the Windows native Mingw cross compiler to Spack, but we'd still need some way to at least run make/autotools and at most port more things, which a non trivial effort to support a very small use case, and at that point we're recreating Cygwin/Msys, which we should not be doing for a multitude of reasons.

> how did obtain binaries then?

Cross compiling for Windows on via the mingw-w64 cross compiler on Linux. The link to the repo that does that is in the new yaml.

alalazo commented:
@johnwparent Why aren't we creating two packages that download those binaries, and rely on the usual bootstrapping procedure?

alalazo reviewed (changes requested):
In view of the meeting today, my feeling is that we don't need a new YAML file and a new bootstrapping method to bootstrap _part_ of Windows[^1]. Reasons for that:

1. Spack can install binary packages too, so we can create e.g. `gpg4win` that downloads the binary that are now at https://spack.github.io/windows-bootstrap-resources/resources/gpg/gpg4win.tar.gz 
2. There seems to be nothing really configurable in the new configuration file. Thus, configuration seems the wrong place for that information.

I understand that cross-compiling is a pain currently, so there's nothing wrong from my point of view in using binaries generated in the other repo. We should document how these binaries are generated, and what makes it complicated to do otherwise though (see https://github.com/spack/windows-bootstrap-resources/issues/5).

[^1]: `clingo` follows a different path

scheibelp commented:
> my feeling is that we don't need a new YAML file

I am not saying this PR is the only possible way to get the behavior that we want. I think it is a reasonable approach though (and I think that in general we shouldn't have to prove that an implementation is the only way to do things for it to be acceptable).

>> Addressing this with packages creates some awkwardness: these packages don't have real versions, and are irrelevant outside of usage within core spack (presumably we'd want a separate package repo for them)

> This is how we do bootstrapping since we started bootstrapping clingo. We have a package clingo-bootstrap which is just clingo with some options/tweaks that are relevant to Spack only

`clingo-bootstrap` is a source-based build of `clingo` with extra options and patching. The `file`/`gpg` utilities installed here are binaries. Also, when I say they "don't have real versions" I mean they don't have *any* meaningful version (which makes them different from other packages which install binaries directly): we'd be adding a version just to fit the packaging model.

As far as `clingo-bootstrap` being in the `builtin` repo: that does indeed create a precedent; I think it would be better to have it in a separate repo though (I don't think an end user would ever want to install it).

> Also, doesn't Windows bootstrap clingo from sources, with the usual mechanism? If that's the case Windows would need 2 different mechanisms and code paths for two different dependencies of Spack. This would be inconsistent.

Yes. Currently all platforms have 2 mechanisms: binary bootstrapping and source bootstrapping. This adds a 3rd mechanism with distinct advantages for low-level tools. I wouldn't call it inconsistent because the contexts for the usage of each mechanism differ.

> There seems to be nothing really configurable in the new configuration file. Thus, configuration seems the wrong place for that information.

This is... true, but in some sense seems irrelevant: using Spack's config is convenient for validation (via a schema) but not essential. You're right that this doesn't make use of any other feature of configs (like merging, or scopes). But:

* I think we have established uses of configuration that don't make use of all the features of config (e.g. ci and environments)
* Overall I assume this complaint is trivial to address: we can probably do validation of a yaml file without accessing the higher-level machinery of `config.py`

> I don't understand what you mean by 2-layer bootstrapping.

Apologies on my end: my argument here was weak. Overall my suspicion is that a package-based approach would also involve modifications to core bootstrapping code (not just adding more specs to `bootstrap/core`). I'd rather focus on the benefits of the current approach.

alalazo commented:
> Also, when I say they "don't have real versions" I mean they don't have any meaningful version (which makes them different from other packages which install binaries directly): we'd be adding a version just to fit the packaging model.

`clingo-bootstrap` has a `@spack` version pinning a commit, which seems pretty similar to what's needed here. 

The packages in this PR would install from binaries because of https://github.com/spack/spack/pull/41810#issuecomment-2214199959, and that's completely fine in my opinion. In the docstring of each package, we can describe its use, and point at https://github.com/spack/windows-bootstrap-resources in the homepage. I don't really see what we are missing, if we use packages and leverage packages installation procedures (instead of adding a new code path to perform an installation).

> This is... true, but in some sense seems irrelevant: using Spack's config is convenient for validation (via a schema) but not essential. 

Adding new configuration, exposed to users, because it's convenient for schema validation, is in my opinion a bad argument. If we used a package, validation would occur using the sha256 of the version, the `url` would be in the package. No need to validate a schema.

johnwparent committed:
Alternative Package approach

scheibelp commented on var/spack/repos/builtin/packages/win-gpg/package.py:
```diff
@@ -0,0 +1,39 @@
+
+
+    def install(self, spec, prefix):
+        shutil.move(self.stage.source_path, prefix)
```
(question) The source path ends with `spack-src`: is this intending to move everything inside `spack-src` to the prefix?

If the prefix already exists, it might end up with `spack-src` inside of it. I don't think we create the prefix before the install, so this should work, but might be more robust to explicitly move the things inside `source_path`

johnwparent replied:
> is this intending to move everything inside spack-src to the prefix

Yes, the contents of `spack-src` should just be install tree. 
I'll refactor to move all the subfolders.

scheibelp commented on var/spack/repos/builtin/packages/win-file/package.py:
```diff
@@ -0,0 +1,36 @@
+    """
+
+
+    homepage = "https://www.darwinsys.com/file/"
```
IMO the Github page where we host this would be more useful

johnwparent replied:
Ok I was kind of flip-flopping on this one, so thanks for weighing in. I wanted to be sure to leave a link to the original project to make it clear that this is actually file, just cross compiled. 

I think best option would be to link to the GH repo here, and have the GH repo readme, and pages instance provide more details (which we can hash out on that repo since all we need to update here is this).

scheibelp commented on var/spack/repos/builtin/packages/file/package.py:
```diff
@@ -38,6 +38,8 @@ class File(AutotoolsPackage):
     depends_on("zlib-api")
     depends_on("zstd", when="@5.44:")
 
+    provides("file")
```
IMO we should explicitly request `win-file` and `win-gpg` vs. creating a virtual. 

* Since `file` doesn't work on Windows, and `win-file` doesn't work on not-Windows, there is limited value in presenting these as virtuals
* I'm not actually sure what happens when a provider name matches a virtual name but we generally avoid it

johnwparent replied:
> I'm not actually sure what happens when a provider name matches a virtual name but we generally avoid it

This was also a concern I had but it didn't _seem_ to break anything?


So the reason I flipped to @alalazo's position on this PR was because I am pretty constantly thinking about how we can offer packages that don't build with Windows native tools (or require unix headers/libs), but can be compiled for the platform (Conan and Vcpkg both offer "unix only" packages via prebuilt binaries, similar to what we do here). Since we're already building these packages, it made sense to me that we offer them as first class packages (basically Max's suggestion). However, I felt having to differentiate "file" vs "win-file" across platforms/contexts really limited the utility of even offering `win-file` as a first class package beyond bootstrapping, as it requires uses to think about what platform they're on to select a package name, and makes envs less distributable/easily cross platform. Ideally the users would just request "file" and depending on the platform they're on, they would be served the correct package. AFAIK, virtuals are the only way to do this

scheibelp replied:
>> I'm not actually sure what happens when a provider name matches a virtual name but we generally avoid it

> This was also a concern I had but it didn't seem to break anything?

I think even if it works, it's a good policy to use separate names because otherwise the error messages can be confusing. If we wanted to go the route of allowing package names to shadow virtual names, I think we should also explicitly test for it (but overall I think that should be a separate PR, so if we keep a virtual, I think that `file` and `win-file` should both provide something new for now, like `file-util`).

> I felt having to differentiate "file" vs "win-file" across platforms/contexts really limited the utility of even offering win-file as a first class package beyond bootstrapping, 

True. However, this is an atypical usage of virtuals: typically multiple providers for a virtual would be expected to be interchangeable but as stated above, the `win-file` package would not be desired/usable on non-Windows (would it?); that makes me want to take a step back and consider how the model could accommodate users trying to install `file` functionality on Windows.

To clarify, this is not an issue of lacking constructs in Spack - you can say `if sys.platform == "win32": provides(...)` - but more about whether handling this use case by introducing a non-idiomatic provider is the way to go.

scheibelp reviewed (changes requested):
I have a few comments about arrangement but overall this looks good

I presume (if this switchover completes)

* Most files will be removed from the diff (including added functionality like `binary_resource`, the `bootstrap_resource` schema
* A call to `ensure_executables_in_path_or_raise` will be added (similar to `ensure_gpg_in_path_or_raise`, but for `win-file`

johnwparent committed:
style

johnwparent committed:
Conditional virtuals

johnwparent committed:
Full switchover to new resource bootstrapping

johnwparent committed:
prevent bad imports

johnwparent committed:
style

johnwparent committed:
style

johnwparent committed:
Proper rename of file

scheibelp commented on var/spack/repos/builtin/packages/nixfile/package.py:
```diff
@@ -4,11 +4,12 @@
 
 
-class File(AutotoolsPackage):
+class NixFile(AutotoolsPackage):
```
`nixfile` <--> `Nixfile`, `nix-file` <--> `NixFile` (this is why audit is failing)

That being said, making what was once a package name into a virtual has the potential to disrupt user caches. IMO it would be a smoother transition to make the virtual `file-util` and have `file/win-file` provide `file-util` (users on windows who select file can follow the provider info to get the relevant provider of `file-util` on their system).

johnwparent replied:
Yeah that's a better idea, will do.

scheibelp commented on lib/spack/spack/util/gpg.py:
```diff
@@ -63,7 +63,7 @@ def init(gnupghome=None, force=False):
     with spack.bootstrap.ensure_bootstrap_configuration():
         spack.bootstrap.ensure_gpg_in_path_or_raise()
-        GPG, GPGCONF = _gpg(), _gpgconf()
+    GPG, GPGCONF = _gpg(), _gpgconf()
```
(question) Were you seeing a problem with this when it was inside the context manager? Overall, after reviewing uses of `ensure_bootstrap_configuration`, they seem inconsistent to me, so I'm not actually sure which is strictly correct. Generally, the original organization might be more robust:

1. `ensure_bootstrap_configuration` disables any active spack envs,
2. For now, re/activating the `Environment` object does not modify `PATH` the same way that `spack env activate` does (so what you did should generally be OK)
3. `ensure_gpg_in_path_or_raise` will prepend to PATH if an installation occurs as part of meeting its guarantee

if [2] changes, this change could be problematic (i.e. would interfere with [3]). I also wouldn't have expected the original organization to cause you any trouble.

johnwparent replied:
Oh I actually thought this _was_ restoring it to previous behavior, I couldn't remember if I had initially moved it under the context manager during the original approach. Should have checked the diff, will restore.

johnwparent committed:
Review

johnwparent committed:
Give virtual provider defaults

johnwparent committed:
Expose bootstrapping api

johnwparent commented:
@scheibelp I think the current concretizer behavior requires a deeper understanding than I currently have, is this due to the fact gpg is now a virtual or the fact we're not concretizing with Clingo? Either way, could use some suggestions to more forward, thanks!

scheibelp commented:
> is this due to the fact gpg is now a virtual 

I think this is the case

> or the fact we're not concretizing with Clingo?

Recently we switched to using clingo everywhere, so perhaps the opposite. From testing back to Jan. 1 2024 we can't do `spack spec virtual%compiler`.

The root spec logic has been attaching compilers for a while, so I'm not sure what changed recently to make the CI check fail (I would have thought this problem would occur as soon as `gnupg` root spec had been replaced with a virtual)

For now, we may need the `gpg_root_spec` function to branch on system type

johnwparent committed:
Condition gpg_root_spec per platform

johnwparent commented:
> The root spec logic has been attaching compilers for a while, so I'm not sure what changed recently to make the CI check fail (I would have thought this problem would occur as soon as `gnupg` root spec had been replaced with a virtual)
> 
> For now, we may need the `gpg_root_spec` function to branch on system type

Makes sense, I figured as much but was really hoping we'd get to keep this nice and seamless. Oh well, pushed a the required changes!

scheibelp commented:
@spackbot run pipeline

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -465,14 +465,20 @@ def clingo_root_spec() -> str:
 
+def file_root_spec() -> str:
+    """Return the root spec used to bootstrap file"""
+    return _root_spec("file")
```
Should this be `"win-file" if IS_WINDOWS else "file"`?

johnwparent replied:
Right we should do it for both, good catch.

johnwparent committed:
use specific file name too

scheibelp commented:
@spackbot run pipeline

johnwparent commented:
@scheibelp I think this one is good to go now?

alalazo reviewed (changes requested):
Basically LGTM, thanks! I have a few minor change requests, mainly to polish the PR:

1. The two new virtuals added here are not used anymore (they were used initially in `spack.bootstrap.core`), let's just remove them
2. We miss a test to bootstrap file (on all systems) and one to bootstrap win-gpg.[^1]
3. We need a small update so that `spack bootstrap status` shows "file" as something we can bootstrap
4. The "only_windows" marker is not used, and can be removed. We can add it back in another PR
5. (Nitpick) I would just rename "spack_filesystem.py" to "filesystem.py", and use regular imports

I implemented most of these changes in https://github.com/johnwparent/spack/pull/11, feel free to rebase and merge the PR if you agree with them. The only thing missing is point number 2 above.

[^1]:  For Windows we might want to turn https://github.com/spack/spack/actions/runs/10458707326/job/28961322763?pr=41810 into something that does `spack bootstrap now`

johnwparent commented:
2, 3, and 5, Sounds good! 

Wr.t. 1:...
> The two new virtuals added here are not used anymore (they were used initially in spack.bootstrap.core), let's just remove them

While they're no longer directly used by Spack core, they are still packages in Spack exposing different "implementations" (i.e. a Windows port) of the same functionality, like other virtuals. This is how we've been handling packages that are sufficiently disparate across platforms to the point where they require a separate package recip to this point (i.e. bison/flex among others).
Otherwise packages depending on file for some reason need to specify one dep per platform and users have to track which file they want available. This makes environments less useable cross platform.


> (Nitpick) I would just rename "spack_filesystem.py" to "filesystem.py", and use regular imports

Would this not be confusing due to `llnl.util.filesystem`? Obviously the modules differentiate them, but wanting to avoid confusing other devs with two `filesystem.py`'s is the reason I chose the `spack_filesystem` name. I don't really have a strong opinion, just wanted to raise that concern.

alalazo commented:
> Otherwise packages depending on file for some reason need to specify one dep per platform and users have to track which file they want available. This makes environments less useable cross platform among other UX concerns.

To be clear, I'm not saying to never add the virtuals, just to not add them here - since we don't have a use case for that. If in a following PR we add, say, `file-api` and change a few packages to use it then that is an absolutely acceptable change.

johnwparent commented:
> To be clear, I'm not saying to never add the virtuals, just to not add them here

Ok that makes more sense, I was very confused as to why the virtuals themselves were an issue, thank you for elaborating

 > since we don't have a use case for that.

Is a user installation not a sufficient use case? Or environments? I'm just trying to understand the rationale behind requiring such a concerete, "in-Spack" use case, or why we'd delay adding this virtual when we're adding the package here?

johnwparent committed:
Review

johnwparent committed:
style

johnwparent committed:
Add test for bootstrap

johnwparent commented:
@scheibelp @alalazo changes have been made, please re-review as is convenient! Thanks

alalazo reviewed (changes requested):
> Is a user installation not a sufficient use case? Or environments? I'm just trying to understand the rationale behind requiring such a concerete, "in-Spack" use case, or why we'd delay adding this virtual when we're adding the package here?

Rationale is very simple:
- The PR is about bootstrapping gpg and file on Windows
- It adds 2 unrelated virtuals, which are not needed for the scope of the PR, nor are used by any package
- $\implies$ They don't belong to the PR 

For your convenience you can cherry-pick f877ad5298be7a43241e3cf15284571ceab4fbbd to remove them. 

The new test is currently failing, see https://github.com/spack/spack/actions/runs/10514727437/job/29134663321?pr=41810

johnwparent committed:
Fix test

johnwparent committed:
remove virtuls so they can be added back

johnwparent committed:
unused imports

johnwparent commented:
@scheibelp changes have been made please review when convenient! Thanks

johnwparent committed:
Bootstrap file

alalazo reviewed (changes requested):
Did anybody check the test results? Currently the new tests are passing because the runner images have the software we want to bootstrap [already installed](https://github.com/spack/spack/actions/runs/10529679345/job/29179159531):
![2024-08-26_08-27](https://github.com/user-attachments/assets/1b6d2f72-9009-43f9-bef6-ffba5596c893)


For other platforms we do:

https://github.com/spack/spack/blob/eccecba39ac681ebf1f78cd37fe0409f0b966e05/.github/workflows/bootstrap.yml#L117-L126


to test GnuPG bootstrapping. We may need something similar for Windows. 

Also, we are still not testing `file` bootstrapping on `linux`. We might want to use a Docker image, since that is likely the main use case for bootstrapping `file` on `linux`.


OT: @johnwparent For my understanding of the workflow here. The small changes proposed in https://github.com/johnwparent/spack/pull/11 have been incorporated here almost verbatim, so why closing the PR and redoing the commits (feels like duplicated work)?

scheibelp commented:
Your first point about this not actually exercising the bootstrap mechanism on Windows is good. However:

> Also, we are still not testing file bootstrapping on linux. We might want to use a Docker image, since that is likely the main use case for bootstrapping file on linux.

This seems redundant: on Linux we are already testing bootstrapping of one thing (gpg) - why should we be explicitly testing this same mechanism for each thing that could be bootstrapped?

scheibelp commented:
> The small changes proposed in https://github.com/johnwparent/spack/pull/11 have been incorporated here almost verbatim, so why closing the PR and redoing the commits (feels like duplicated work)?

Also @alalazo is your concern about co-authorship?

I do this myself often because I want to very closely understand all the edits the person is suggesting.

alalazo commented:
> Also @alalazo is your concern about co-authorship?

No, it's about the workflow. I tried to speed-up the merge of this PR by providing the trivial stuff I noticed, and pointing to the missing test - hoping that this will leave just one thing to focus on. It seems this attempt of mine failed.

alalazo commented:
> This seems redundant: on Linux we are already testing bootstrapping of one thing (gpg) - why should we be explicitly testing this same mechanism for each thing that could be bootstrapped?

This PR makes `file` bootstrappable on `linux`, which is great. I think we should just add something that will raise a flag if that stops to work, like we do for any dependency that we can currently bootstrap.

scheibelp commented:
>> This seems redundant: on Linux we are already testing bootstrapping of one thing (gpg) - why should we be explicitly testing this same mechanism for each thing that could be bootstrapped?

> This PR makes file bootstrappable on linux, which is great. I think we should just add something that will raise a flag if that stops to work, like we do for any dependency that we can currently bootstrap.

If the same mechanism is used for both, I don't see why each needs to be tested individually.

I assume this can be also be resolved by only bootstrapping `file` on Windows.

alalazo commented:
> If the same mechanism is used for both, I don't see why each needs to be tested individually.

For bootstrapping we are not testing the mechanism only, but that each item that can be bootstrapped is bootstrapped correctly (see the jobs, we have multiple for `clingo` and `gnupg`). 

If it makes sense we can join some specs in the same job, but definitely we want to be sure that we don't end in the situation where the general mechanism is working, but `gnupg` etc. stops being bootstrappable for some reason.

johnwparent commented:
> Did anybody check the test results? Currently the new tests are passing because the runner images have the software we want to bootstrap [already installed](https://github.com/spack/spack/actions/runs/10529679345/job/29179159531)

I only checked that they didn't fail, didn't even occur to me that gpg4win would be in these GH runners. Nice catch.
FWIW we can know this works because Gitlab CI is passing, and we're doing this bootstrapping there (this is also the motivating use case), but we should have an explicit check that actually does the bootstrapping, I'll fix this thanks.

> For other platforms we do:
> 
> https://github.com/spack/spack/blob/eccecba39ac681ebf1f78cd37fe0409f0b966e05/.github/workflows/bootstrap.yml#L117-L126
> 
> to test GnuPG bootstrapping. We may need something similar for Windows.
>

Thanks for the direction, will implement the Windows equivalent. 

> Also, we are still not testing `file` bootstrapping on `linux`. We might want to use a Docker image, since that is likely the main use case for bootstrapping `file` on `linux`.

Did you have in mind something similar to the workflow we have to produce the binary caches for gpg?


> OT: @johnwparent For my understanding of the workflow here. The small changes proposed in [johnwparent#11](https://github.com/johnwparent/spack/pull/11) have been incorporated here almost verbatim, so why closing the PR and redoing the commits (feels like duplicated work)?

My apologies if I've stepped on your toes at all, it's more just a workflow preference on my part.

Basically I a) almost never just accept code suggestions made in the comment tab (or otherwise) unless they're incredibly trivial, for numerous reasons, but mostly as a form of review for the suggestion so I don't propagate mistakes but also because it forces me to actually consider what the edits are doing, and b) making the edits myself forces me to reason about why they are necessary, my initial approach, and further improvements I could make/other places I can apply general purpose suggestions.

In this specific case, a lot of the changes you made were removing code that would eventually be intended for another PR, so rather than have to replicate those changes by hand, I reworked by local history so I could cherry pick those out into their own PRs and then drop the commits from the history so on my end the "just cherry-pick" option seemed a bit slower/unwieldy. I prefer to leverage git to re-coding. I appreciate the intent behind the effort. 

> I tried to speed-up the merge of this PR by providing the trivial stuff I noticed, and pointing to the missing test - hoping that this will leave just one thing to focus on. It seems this attempt of mine failed.

Frankly I spent more time thinking about the virtuals change than I did adding the test LOL.

>For bootstrapping we are not testing the mechanism only, but that each item that can be bootstrapped is bootstrapped correctly (see the jobs, we have multiple for clingo and gnupg).

@scheibelp FWIW I agree with @alalazo here, after maintaining Windows package support without explicit CI for a long time, for mission critical things like bootstrapping, having the explicit test is the only way to be genuinely confident the build still works. Maybe `file` is sufficiently static and simple that it will realistically never fail, but I thought that about `cpuinfo` and `abseil-cpp` and both have been broken on Windows w/o CI multiple times in the past.

johnwparent committed:
Purge gpg on Windows

johnwparent commented:
@scheibelp @alalazo pushed up changes removing Gpg on Windows (works locally, untested on CI, we'll see if we have permissions to do this, if not we may just have to rely on Gitlab to test this).
Waiting for @alalazo's response w.r.t. the linux `file` bootstrap test to determine path forward.

johnwparent committed:
Windows need from "source" bootstrapping

johnwparent committed:
Correct gpg name

johnwparent committed:
url typo

johnwparent committed:
Correct install

johnwparent committed:
Don't bootstrap file on unixes

johnwparent commented:
@spackbot fix style

alalazo reviewed (commented):
win-gpg seems to be bootstrap, but win-file is probably already on the runner? I don't see it in the store (end of the log for [this job](https://github.com/spack/spack/actions/runs/10604751475/job/29392480768?pr=41810)).

Also (minor, can be done in a later PR), since it seems we are interleaving changes just for Windows, can we refactor the interleaved logic in a new Windows job that tests, in succession, clingo, gpg and file? That would possibly be more readable both on windows, and on macOS / linux

alalazo commented:
@johnwparent You edited my review instead of replying in https://github.com/spack/spack/pull/41810#pullrequestreview-2271357921 Quite confusing for a future reader :slightly_frowning_face:

johnwparent commented:
> @johnwparent You edited my review instead of replying in [#41810 (review)](https://github.com/spack/spack/pull/41810#pullrequestreview-2271357921) 

So I did! Sorry about that, wasn't intentional, I must have just hit edit instead of "quote reply". I've restored the comment to its original state and left my "edit" as its own comment as was intended.  

> Quite confusing for a future reader 🙁

Forget future readers, it was confusing to me!

johnwparent commented:
@alalazo 

win-gpg seems to be bootstrap, but win-file is probably already on the runner? I don't see it in the store (end of the log for [this job](https://github.com/spack/spack/actions/runs/10604751475/job/29392480768?pr=41810)).

It's definitely not being bootstrapped, that said, I don't see it explicitly listed in the [runner provision manifest](https://github.com/actions/runner-images/blob/main/images/windows/Windows2022-Readme.md).
The runner images do have msys installed, which has previously caused issues like this but I see it's not on the PATH so that's somewhat unlikely. There's also GNU Bin utils, which also might vendor file.

I'm going to add a commit that finds file for us, and then we can remove it for the bootstrap.

FWIW for review purposes, we can confirm the file bootstrap works via the [gitlab job output](https://gitlab.spack.io/spack/spack/-/jobs/12363404#L76)

> Also (minor, can be done in a later PR), since it seems we are interleaving changes just for Windows, can we refactor the interleaved logic in a new Windows job that tests, in succession, clingo, gpg and file? That would possibly be more readable both on windows, and on macOS / linux

Sure, will do in the follow up.

johnwparent committed:
Runner machines have file evidently

johnwparent committed:
Style

johnwparent commented:
@alalazo added the file remove (we'll see if it works/needs more thought behind a naive find and remove). If that passes I believe I'll have addressed all your comments. 

@scheibelp I believe Harmen resolved the Gitlab CI issue we were seeing with Windows at the very least, I've rebased on develop and force pushed, hopefully this will be the end of the CI issues holding this up.

johnwparent committed:
typo

johnwparent committed:
Windows multi proc idiosyncracy

johnwparent committed:
style

alalazo reviewed (approved):
:relieved:

alalazo commented:
The linux test for file is still missing, but we can add it in a following PR. Same for simplifying the CI job as suggested here https://github.com/spack/spack/pull/41810#pullrequestreview-2271357921
